js/src/data/actions.js

```diff
@@ -72,77 +73,80 @@ export function* fetchShippingRates() {
 }
 
 /**
- * Aggregated shipping rate.
+ * Saves shipping rates.
  *
- * @typedef {Object} AggregatedShippingRate
- * @property {Array<CountryCode>} countries Array of destination country codes.
- * @property {string} currency Currency of the price.
- * @property {number} rate Shipping price.
- */
-
-/**
- * Updates or inserts given aggregated shipping rate.
+ * This is done by removing the old shipping rates first,
+ * and then upserting the new shipping rates.
  *
- * @param {AggregatedShippingRate} shippingRate
+ * @param {Array<ShippingRate>} newShippingRates
  */
-export function* upsertShippingRates( shippingRate ) {
-	const { countryCodes, currency, rate } = shippingRate;
-
+export function* saveShippingRates( newShippingRates ) {
 	try {
-		yield apiFetch( {
-			path: `${ API_NAMESPACE }/mc/shipping/rates/batch`,
-			method: 'POST',
-			data: {
-				country_codes: countryCodes,
-				currency,
-				rate,
-			},
-		} );
+		/**
+		 * Compare the new shipping rates and the old ones from the store
+		 * to find out the old ones to be deleted.
+		 */
+		const oldShippingRates = yield select( STORE_KEY, 'getShippingRates' );
+		const deleteIds = oldShippingRates.reduce( ( acc, cur ) => {
+			const found = newShippingRates.find( ( el ) =>
+				isSameShippingRate( el, cur )
+			);
+
+			if ( ! found ) {
+				acc.push( cur.id );
+			}
```

Comment:
❓ 

What do you think if we do something like the code below so it is more descriptive and easier to test?

```suggestion

const isInNewShippingRates = ( item,  newShippingRates = [] ) => {
  return newShippingRates.some( ( el ) =>
  isSameShippingRate( el, item )
  );
}

.......

const deleteIds = oldShippingRates.filter((item) => ! isInNewShippingRates( item, newShippingRates ) ).map( item => item.id);


```
We could also add some tests for this part of the code.

Reply:
I did some changes (see [`18ee6a6` (#1262)](https://github.com/woocommerce/google-listings-and-ads/pull/1262/commits/18ee6a62df8acb20d7f064cdd72a7c5bc2d506b1)) based on your suggestion, and I do think they are much easier to read now. I have also added tests for it. Thanks for the idea! 🙂  👍